File: app/events/schemas/events.py
from typing import Optional, List, Union
from pydantic import BaseModel
from datetime import datetime
from typing import Literal
from enum import Enum

class EventType(str, Enum):
    PAGEVIEW = "pageview"
    CLICK = "click"
    SCROLL = "scroll"
    MEDIA = "media"
    FORM = "form"
    CONVERSION = "conversion"
    ERROR = "error"
    VISIBILITY = "visibility"
    PERFORMANCE = "performance"
    CUSTOM = "custom"

class BaseEvent(BaseModel):
    globe_id: str
    event_id: str
    timestamp: datetime
    session_id: str
    client_timestamp: datetime
    event_type: str

class PageView(BaseEvent):
    url: str
    referrer: Optional[str]
    title: str
    path: str
    viewport_width: int
    viewport_height: int
    load_time: float

class Click(BaseEvent):
    element_path: str
    element_text: str
    x_pos: int
    y_pos: int
    href: Optional[str]

class Scroll(BaseEvent):
    depth: int
    direction: str
    max_depth: int
    relative_depth: float

class Media(BaseEvent):
    media_type: Literal["video", "audio"]
    action: Literal["play", "pause", "complete"]
    media_url: str
    playback_time: int
    duration: int
    title: Optional[str]

class Form(BaseEvent):
    form_id: str
    action: str
    fields: List[str]
    success: bool
    error_message: Optional[str]

class Conversion(BaseEvent):
    conversion_type: str
    value: float
    currency: str
    products: Optional[List[str]]

class Error(BaseEvent):
    error_type: str
    message: str
    stack_trace: str
    component: str

class Visibility(BaseEvent):
    visibility_state: str

class Performance(BaseEvent):
    metric_name: str
    value: float
    navigation_type: str
    effective_connection_type: str

# Union type for all possible event types
Event = Union[
    PageView,
    Click, 
    Scroll,
    Media,
    Form,
    Conversion,
    Error,
    Performance,
    Visibility
]

# Mapping of event_type strings to event classes
EVENT_TYPE_MAPPING = {
    "pageview": PageView,
    "click": Click,
    "scroll": Scroll,
    "media": Media,
    "form": Form,
    "conversion": Conversion,
    "error": Error,
    "performance": Performance,
    "visibility": Visibility
}

File: static/ts/types/events.ts
import { z } from "zod";

// Base event type that all others extend from
const BaseEvent = z.object({
  event_id: z.string(),
  timestamp: z.date(),
  session_id: z.string(),
  user_id: z.string(),
  client_timestamp: z.date(),
  event_type: z.string(),
});

const PageViewEvent = BaseEvent.extend({
  event_type: z.literal("pageview"),
  data: z.object({
    url: z.string(),
    referrer: z.string().optional(),
    title: z.string(),
    path: z.string(),
    viewport: z.object({
      width: z.number(),
      height: z.number(),
    }),
    load_time: z.number(),
  }),
});

const ClickEvent = BaseEvent.extend({
  event_type: z.literal("click"),
  data: z.object({
    element_path: z.string(),
    element_text: z.string(),
    x_pos: z.number(),
    y_pos: z.number(),
    href: z.string().optional(),
  }),
});

const ScrollEvent = BaseEvent.extend({
  event_type: z.literal("scroll"),
  data: z.object({
    depth: z.number(),
    direction: z.string(),
    max_depth: z.number(),
    relative_depth: z.number(),
  }),
});

const MediaEvent = BaseEvent.extend({
  event_type: z.literal("media"),
  data: z.object({
    media_type: z.enum(["video", "audio"]),
    action: z.string(),
    media_url: z.string(),
    playback_time: z.number(),
    duration: z.number(),
    title: z.string().optional(),
  }),
});

const FormEvent = BaseEvent.extend({
  event_type: z.literal("form"),
  data: z.object({
    form_id: z.string(),
    action: z.string(),
    fields: z.array(z.string()),
    success: z.boolean(),
    error_message: z.string().optional(),
  }),
});

const ConversionEvent = BaseEvent.extend({
  event_type: z.literal("conversion"),
  data: z.object({
    conversion_type: z.string(),
    value: z.number(),
    currency: z.string(),
    products: z.array(z.string()).optional(),
  }),
});

const ErrorEvent = BaseEvent.extend({
  event_type: z.literal("error"),
  data: z.object({
    error_type: z.string(),
    message: z.string(),
    stack_trace: z.string(),
    component: z.string(),
  }),
});

const PerformanceEvent = BaseEvent.extend({
  event_type: z.literal("performance"),
  data: z.object({
    metric_name: z.string(),
    value: z.number(),
    navigation_type: z.string(),
    effective_connection_type: z.string(),
  }),
});

const CustomEvent = BaseEvent.extend({
  event_type: z.literal("custom"),
  name: z.string(),
  data: z.record(z.unknown()),
});

// Export the combined Event type
export type Event =
  | z.infer<typeof PageViewEvent>
  | z.infer<typeof ClickEvent>
  | z.infer<typeof ScrollEvent>
  | z.infer<typeof MediaEvent>
  | z.infer<typeof FormEvent>
  | z.infer<typeof ConversionEvent>
  | z.infer<typeof ErrorEvent>
  | z.infer<typeof PerformanceEvent>
  | z.infer<typeof CustomEvent>;

export type BatchAnalyticsRequest = {
  events: Event[];
};

// Export the schemas for validation
export const EventSchemas = {
  pageview: PageViewEvent,
  click: ClickEvent,
  scroll: ScrollEvent,
  media: MediaEvent,
  form: FormEvent,
  conversion: ConversionEvent,
  error: ErrorEvent,
  performance: PerformanceEvent,
  custom: CustomEvent,
} as const;
